### PR TITLE
fix RemoteMergeOperation to always use an up-to-date remote-base

### LINKED ISSUE #371: pull-request merge making changes not made in PR
Made a test repo https://github.com/aerialship/test
Forked it to https://github.com/tmilos/test 

Made first test PR https://github.com/aerialship/test/pull/2 and it produced merged changes https://github.com/aerialship/test/commit/8cea6dc5b002ea80275ab48aea3addf24ba1885e ... All well

Problematic is next PR https://github.com/aerialship/test/pull/3/files with changes on only one file, but after ```gush pull:merge --squash 3``` it makes changes on 3 files https://github.com/aerialship/test/commit/15b0e158d848f3b74f2545e01f296c09be3d9078

Here's the log

```
d:\www\home\aerial\test>gush pull:merge --squash 3 -vvv
CMD "git" "config" "--local" "--get" "remote.aerialship.url"
CMD "git" "config" "--local" "--get" "remote.aerialship.url"

[INFO] Adding remote 'aerialship' with 'git@github.com:aerialship/test.git' to git local config.
CMD "git" "config" "--local" "--get" "remote.aerialship.url"
CMD "git" "config" "--local" "--get" "remote.aerialship.url"
CMD "git" "config" "remote.aerialship.url" "https://github.com/aerialship/test.git" "--local"
CMD "git" "config" "--local" "--get" "remote.aerialship.pushurl"
CMD "git" "config" "remote.aerialship.pushurl" "git@github.com:aerialship/test.git" "--local"
CMD "git" "config" "--local" "--get" "remote.tmilos.url"
CMD "git" "config" "--local" "--get" "remote.tmilos.url"

[INFO] Adding remote 'tmilos' with 'git@github.com:tmilos/test.git' to git local config.
CMD "git" "config" "--local" "--get" "remote.tmilos.url"
CMD "git" "config" "--local" "--get" "remote.tmilos.url"
CMD "git" "config" "remote.tmilos.url" "https://github.com/tmilos/test.git" "--local"
CMD "git" "config" "--local" "--get" "remote.tmilos.pushurl"
CMD "git" "config" "remote.tmilos.pushurl" "git@github.com:tmilos/test.git" "--local"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "remote" "update" "aerialship"
CMD "git" "reset" "--hard" "HEAD~1"
CMD "git" "pull" "-u" "aerialship" "master"
CMD "git" "remote" "update" "tmilos"
CMD "git" "checkout" "tmilos/bar"
CMD "git" "checkout" "-b" "tmp--tmilos--bar"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "checkout" "tmp--tmilos--bar"
CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%ae" "master..tmp--tmilos--bar"
CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%H" "--reverse" "master..tmp--tmilos--bar"
CMD "git" "--no-pager" "show" "--format=%an <%ae>%n%s%n%n%b" "--no-color" "--no-patch" "031aa4a219758134cb1bea02b7cc9ae393cd9af8"
CMD "git" "reset" "--soft" "master"
CMD "git" "commit" "-F" "C:\Users\mtomic\AppData\Local\Temp\gush\571E.tmp" "-a" "--author" "Milos Tomic <tmilos@gmail.com>"
CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%H" "--reverse" "master..tmp--tmilos--bar"
CMD "git" "--no-pager" "show" "--format=%an <%ae>%n%s%n%b" "--no-color" "--no-patch" "110e31b813faeef9f4a4784d5e4d116ad5965b9d"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "checkout" "master"
CMD "git" "merge" "--no-ff" "--no-commit" "tmp--tmilos--bar"
OUT Removing doc/foo.md
OUT Automatic merge went well; stopped before committing as requested
CMD "git" "commit" "-F" "C:\Users\mtomic\AppData\Local\Temp\gush\5804.tmp"
OUT [master 15b0e15] merge #3 Bar (tmilos)
CMD "git" "rev-parse" "HEAD"
CMD "git" "push" "aerialship" "master"
CMD "git" "remote" "update" "aerialship"
CMD "git" "notes" "--ref=github-comments" "add" "-F" "C:\Users\mtomic\AppData\Local\Temp\gush\94B3.tmp" "15b0e158d848f3b74f2545e01
f296c09be3d9078"
CMD "git" "push" "aerialship" "refs/notes/github-comments"
This PR was squashed before being merged into the master branch (closes #3).
CMD "git" "branch" "-D" "tmp--tmilos--bar"

d:\www\home\aerial\test>
```
gush 1.14.1, php 5.4.19, windows 7 x64

Just to be sure, tried it three times :) but always same results

### PR DESCRIPTION
**Pull request by sstok**
|Q            |A   |
|---          |--- |
|Fixed Tickets|#371|
|License      |MIT |
                    
This fixes #371 and some minor bugs;

- use the correct remote base when merging (not the local branch)
- fix not being able to squash when `--force-squash` option was provided
- restore original branch after merge operation.

### CODE REVIEW

**Comment by sstok** [comment]
In the background this is done:

```
$ ../gush/bin/gush pull:merge -vvv 27
CMD "git" "config" "--local" "--get" "remote.sstok.url"
 ! [NOTE] Adding remote "sstok" with "git@github.com:sstok/gush-experiments-sand

CMD "git" "config" "--local" "--get" "remote.sstok.url"
CMD "git" "remote" "add" "sstok" "git@github.com:sstok/gush-experiments-sandbox.
CMD "git" "config" "--local" "--get" "remote.sstok.pushurl"
CMD "git" "config" "remote.sstok.pushurl" "git@github.com:sstok/gush-experiments
CMD "git" "config" "--local" "--get" "remote.sstok.url"
CMD "git" "config" "--local" "--get" "remote.sstok.url"
 Choose the type of the pull request: :
  [0] feature
  [1] bugfix
  [2] refactor
  [3] tests
  [4] remove
  [5] style
  [6] documentation
  [7] security
 > 1

CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "remote" "update" "sstok"
CMD "git" "remote" "update" "sstok"
CMD "git" "checkout" "sstok/feat-gush"
CMD "git" "checkout" "-b" "tmp--sstok--feat-gush"
CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%H" "--reverse" "master..tmp--sstok--feat-gush"
CMD "git" "--no-pager" "show" "--format=%an <%ae>%n%s%n%b" "--no-color" "--no-patch" "1e9bfbf0b7fc3bf79720f05563321b61ed822d22"
CMD "git" "checkout" "sstok/master"
CMD "git" "checkout" "-b" "tmp--sstok--master"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "checkout" "tmp--sstok--master"
CMD "git" "merge" "--no-ff" "--no-commit" "tmp--sstok--feat-gush"
OUT Automatic merge went well; stopped before committing as requested
CMD "git" "commit" "-F" "C:\Users\Sebastiaan Stok\AppData\Local\Temp\gush\B898.tmp"
OUT [tmp--sstok--master d157ed1] bugfix #27 add .gush.yml (sstok)
CMD "git" "rev-parse" "HEAD"
CMD "git" "push" "sstok" "tmp--sstok--master:master"

 [OK] This PR was merged into the master branch.


CMD "git" "branch" "-D" "tmp--sstok--feat-gush"
CMD "git" "branch" "-D" "tmp--sstok--master"
```

**Comment by sstok** [comment]
He :flushed: still one minor bug
`CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%H" "--reverse" "master..tmp--sstok--feat-gush"` this will uses the old base :scream:

**Comment by sstok** [comment]
And then I discovered another (potential) problem with the `--switch` option, internally this uses `git rebase --onto` but this operation can only work safely when the base-reference is exactly the one that was used!

You can't simply use the original target as that is not the actual branch, it can work but I'm not sure and would rather not take any risks on that.

**Comment by sstok** [comment]
Full operation log,I double checked and it should be good :+1: 

```
$ ../gush/bin/gush pull:merge --switch=master --squash -vvv 29
CMD "git" "config" "--local" "--get" "remote.sstok.url"
CMD "git" "config" "--local" "--get" "remote.sstok.url"
CMD "git" "config" "--local" "--get" "remote.sstok.url"
CMD "git" "config" "--local" "--get" "remote.sstok.url"
 Choose the type of the pull request: :
  [0] feature
  [1] bugfix
  [2] refactor
  [3] tests
  [4] remove
  [5] style
  [6] documentation
  [7] security
 > 6

CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "remote" "update" "sstok"
CMD "git" "remote" "update" "sstok"
CMD "git" "checkout" "sstok/master"
CMD "git" "checkout" "-b" "tmp--sstok--master"
CMD "git" "checkout" "sstok/readme-bla"
CMD "git" "checkout" "-b" "tmp--sstok--readme-bla"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "checkout" "tmp--sstok--readme-bla"
CMD "git" "rebase" "--onto" "tmp--sstok--master" "sstok/master" "tmp--sstok--readme-bla"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "checkout" "tmp--sstok--readme-bla"
CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%ae" "tmp--sstok--master..tmp--sstok--readme-bla"
CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%H" "--reverse" "tmp--sstok--master..tmp--sstok--readme-bla"
CMD "git" "--no-pager" "show" "--format=%an <%ae>%n%s%n%n%b" "--no-color" "--no-patch" "c8a3091f8f2db16f9a670270bbdadcb8216b8ae5"
CMD "git" "reset" "--soft" "tmp--sstok--master"
CMD "git" "commit" "-F" "C:\Users\Sebastiaan Stok\AppData\Local\Temp\gush\2F56.tmp" "-a" "--author" "Sebastiaan Stok <s.stok@rollerscapes.net>"
CMD "git" "--no-pager" "log" "--oneline" "--no-color" "--format=%H" "--reverse" "tmp--sstok--master..tmp--sstok--readme-bla"
CMD "git" "--no-pager" "show" "--format=%an <%ae>%n%s%n%b" "--no-color" "--no-patch" "8d0ecbec9fc612a738cffdaebb8c3766f1568e1f"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "rev-parse" "--abbrev-ref" "HEAD"
CMD "git" "checkout" "tmp--sstok--master"
CMD "git" "merge" "--no-ff" "--no-commit" "tmp--sstok--readme-bla"
OUT Automatic merge went well; stopped before committing as requested
CMD "git" "commit" "-F" "C:\Users\Sebastiaan Stok\AppData\Local\Temp\gush\3099.tmp"
OUT [tmp--sstok--master 479af51] documentation #29 update README.md (sstok)
CMD "git" "rev-parse" "HEAD"
CMD "git" "status" "--porcelain" "--untracked-files=no"
CMD "git" "checkout" "master"
CMD "git" "push" "sstok" "tmp--sstok--master:master"

 [OK] This PR was squashed before being merged into the master branch (closes #29).


CMD "git" "branch" "-D" "tmp--sstok--master"
CMD "git" "branch" "-D" "tmp--sstok--readme-bla"
```

File: src/Operation/RemoteMergeOperation.php [hunk 1]
```diff
@@ -81,39 +80,83 @@ public function performMerge()
...
+            );
         }
 
+        $this->gitHelper->pushToRemote($this->targetRemote, $target);
+    }
+
+    private function createBaseBranch()
+    {
+        $targetBranch = null !== $this->switchBase ? $this->switchBase : $this->targetBranch;
+        $this->targetBase = $this->gitHelper->createTempBranch($this->targetRemote.'--'.$targetBranch);
```
Comment (cordoval) [inline src/Operation/RemoteMergeOperation.php]: what if this temp branch already exists will it be removed and recreated?

Comment (sstok) [inline src/Operation/RemoteMergeOperation.php]: Then the process fails, It only happens when you abort the process before the command-terminate event is triggered.

We can add the microtime for extra randomness :+1:

**Pull request merged**